### PR TITLE
[FLINK-23126] Refactor smoke-e2e into smoke-e2e-common and smoke-e2e-embedded

### PR DESCRIPTION
**Pull request by evans-ye**
# What is the purpose of the change

This JIRA focus on refactoring the existing statefun-smoke-e2e module into:
* statefun-smoke-e2e-common (smoke e2e testing framework such as source, sink, verification, etc)
* statefun-smoke-e2e-embedded (embedded java function)

The end goal of this refactoring is to decouple testing framework with different kinds of functions(embedded, remote Java/Go/Python functions) so that we can have a unified testing framework to test those functions as a whole. Testing other language SDKs will be covered in the umbrella JIRA FLINK-23125.

# Brief change log
1. Let _smoke-e2e-embedded_ module depends on _smoke-e2e-common_ module, and build into fat-jar for runtime e2e testing.
2. In _smoke-e2e-embedded_, protobuf classes are pre-generated since for different kind of SDKs, there might not have proper protobuf gen solution at build time. While _smoke-e2e-common_ remains to be generated at build time.
3. Some naming refinement such as _SimpleProtobufServer_ -> _SimpleVerificationServer_, _source_ -> _command-generator-source_, _sink_ -> _discard-sink_, _verification_ -> _verification-sink_. However, more suitable naming is welcomed.

# Verifying

From project root, do mvn clean install -Prun-e2e-tests and it's passed.

### CODE REVIEW

File: statefun-e2e-tests/statefun-smoke-e2e-embedded/pom.xml [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+                    <artifactId>jackson-annotations</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+
+        <!-- smoke end-to-end common -->
+        <dependency>
+            <groupId>org.apache.flink</groupId>
+            <artifactId>statefun-smoke-e2e-common</artifactId>
+            <version>3.1-SNAPSHOT</version>
```
Comment (tzulitai) [inline statefun-e2e-tests/statefun-smoke-e2e-embedded/pom.xml:92]:
```suggestion
            <version>${project.version}</version>
```

File: statefun-e2e-tests/statefun-smoke-e2e-embedded/pom.xml [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.flink</groupId>
+            <artifactId>statefun-smoke-e2e-common</artifactId>
+            <version>3.1-SNAPSHOT</version>
```
Comment (tzulitai) [inline statefun-e2e-tests/statefun-smoke-e2e-embedded/pom.xml:98]:
```suggestion
            <version>${project.version}</version>
```

File: statefun-e2e-tests/statefun-smoke-e2e-embedded/pom.xml [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+                </exclusion>
+            </exclusions>
+        </dependency>
+
+        <!-- smoke end-to-end common -->
+        <dependency>
+            <groupId>org.apache.flink</groupId>
+            <artifactId>statefun-smoke-e2e-common</artifactId>
+            <version>3.1-SNAPSHOT</version>
+            <scope>compile</scope>
```
Comment (tzulitai) [inline statefun-e2e-tests/statefun-smoke-e2e-embedded/pom.xml:93]:
```suggestion
```

**Review by tzulitai — APPROVED** [review]
LGTM @evans-ye, thanks a lot for the contribution.
The PR is basically ready to merge, minus some nitpicks.

As discussed, please also move the `HarnessTest` class to the embedded smoke E2E module, as that should be a runnable that goes with each individual smoke test.

**Comment by evans-ye** [comment]
Changes Updated:
1. Move _HarnessTest_ to Embedded module and tested by running on IDE.
2. Rename _Fn_ to _CommandInterpreterFn_, rename _SmokeVerificationE2E_ to _SmokeVerificationEmbeddedE2E_
3. Make declared PersistedValue name capital: _state_ -> _STATE_.

**Comment by tzulitai** [comment]
@evans-ye there seems to be a conflict with `master`. Could you do another rebase on `master` and resolve the conflict, and squash all the commits into one? Thank you!

**Comment by evans-ye** [comment]
Yes sure.

**Comment by evans-ye** [comment]
Hmmm. There's a build error:
```
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:testCompile (default-testCompile) on project statefun-smoke-e2e-embedded: Compilation failure
Error:  /home/runner/work/flink-statefun/flink-statefun/statefun-e2e-tests/statefun-smoke-e2e-embedded/src/test/java/org/apache/flink/statefun/e2e/smoke/CommandInterpreterTest.java:[49,24] org.apache.flink.statefun.e2e.smoke.CommandInterpreterTest.MockContext is not abstract and does not override abstract method cancelDelayedMessage(java.lang.String) in org.apache.flink.statefun.sdk.Context
```

Let me try to fix it.

**Comment by evans-ye** [comment]
Thanks for the guidance and review @tzulitai .

**Pull request closed without merging**